ndcube/ndcube.py

```diff
@@ -409,21 +425,64 @@ def extra_coords(self):
 
-        n_dim = len(self.dimensions)
-        if (len(min_coord_values) != len(interval_widths)) or len(min_coord_values) != n_dim:
-            raise ValueError("min_coord_values and interval_widths must have "
-                             "same number of elements as number of data dimensions.")
-        # Convert coords of lower left corner to pixel units.
-        lower_pixels = self.world_to_pixel(*min_coord_values)
-        upper_pixels = self.world_to_pixel(*[min_coord_values[i] + interval_widths[i]
-                                             for i in range(n_dim)])
-        # Round pixel values to nearest integer.
-        lower_pixels = [int(np.rint(l.value)) for l in lower_pixels]
-        upper_pixels = [int(np.rint(u.value)) for u in upper_pixels]
-        item = tuple([slice(lower_pixels[i], upper_pixels[i]) for i in range(n_dim)])
+        n_dim = self.data.ndim
+        # Raising a value error if the arguments have not the same dimensions.
+        # Calculation of upper_corner with the inputing interval_widths
+        # This part of the code will be removed in version 2.0
+        if interval_widths:
+            warnings.warn("interval_widths will be removed from the API in "
+                          "version 2.0, please use upper_corner argument.")
+            if upper_corner:
+                raise ValueError("Only one of interval_widths or upper_corner "
+                                 "can be set. Recommend using upper_corner as "
+                                 "interval_widths is deprecated.")
+            if (len(lower_corner) != len(interval_widths)) or (len(lower_corner) != n_dim):
+                raise ValueError("lower_corner and interval_widths must have "
+                                 "same number of elements as number of data "
+                                 "dimensions.")
+            upper_corner = [lower_corner[i] + interval_widths[i] for i in range(n_dim)]
+        # Raising a value error if the arguments have not the same dimensions.
+        if (len(lower_corner) != len(upper_corner)) or (len(lower_corner) != n_dim):
+            raise ValueError("lower_corner and upper_corner must have same"
+                             "number of elements as number of data dimensions.")
+        if units:
+            # Raising a value error if units have not the data dimensions.
+            if len(units) != n_dim:
+                raise ValueError('units must have same number of elements as '
+                                 'number of data dimensions.')
+            # If inputs are not Quantity objects, they are modified into specified units
+            for i in range(n_dim):
+                if type(lower_corner[i]) is not u.quantity.Quantity:
+                    lower_corner[i] = u.Quantity(lower_corner[i], unit=units[i])
+                if type(upper_corner[i]) is not u.quantity.Quantity:
+                    upper_corner[i] = u.Quantity(upper_corner[i], unit=units[i])
```

Comment:
You don't have to check the types in this for loop.  If the type is already a ```Quantity``` the effect will be to convert the ```Quantity``` to the given unit.  This is probably a more intuitive behaviour if a user gives both ```Quantities``` and sets the ```units``` kwarg.

Reply:
Oh, yes, I didn't think that it is working like that ! I go modify that.